### PR TITLE
Multi-Arch Support

### PR DESCRIPTION
**Pull request by dave-tucker**
Allow libnetwork to be cross compiled on many architectures.
Add dnet binary to the `bin` directory

### CODE REVIEW

**Comment by mrjana** [comment]
@dave-tucker Is this still experimental or can we start reviewing?

Also needs a rebase.

**Comment by dave-tucker** [comment]
Rebasing and testing to see if it's usable yet. Will push shortly

File: netutils/utils_linux.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// +build linux
```
Comment (mrjana) [inline netutils/utils_linux.go]: Now that windows is going to be supported for the daemon and therefore for libnetwork as well, we need a utils_unsupported to atleast to satisfy a build level windows CI task

File: Makefile [hunk 2]
```diff
@@ -22,7 +23,19 @@ build: ${build_image}.created
 	${docker} make build-local
 
 build-local:
-	$(shell which godep) go build -tags experimental ./...
+	@mkdir -p "bin"
+	$(shell which godep) go build -o "bin/dnet-linux-amd64" -tags experimental ./cmd/dnet
+
+cross:
```
Comment (mrjana) [inline Makefile]: For my understanding, this target is not completely validated. Correct?

**Comment by mrjana** [comment]
Other than the addition of an unsupported version of the api the diff LGTM

File: netutils/utils_linux.go (hunk 1 above)
Comment (dave-tucker) [inline netutils/utils_linux.go]: Oddly enough, it builds for me without! I can add one if required though

File: Makefile (hunk 2 above)
Comment (dave-tucker) [inline Makefile]: Ah yes that's a point. I'll do some bash wizardry to fix it.

**Comment by dave-tucker** [comment]
Rebased and ready for review @mrjana. Note that I've now added this to CircleCI so we should catch any issue that break compilation on other platforms. We should still look at expanding CI to cover Windows and FreeBSD though. Either Jenkins and/or [AppVeyor](http://www.appveyor.com/) for Windows

**Comment by mrjana** [comment]
@dave-tucker The builds are failing on this still. Can you please take a look and rebase if necessary?

**Comment by dave-tucker** [comment]
@mrjana yep will take a look today

**Comment by dave-tucker** [comment]
@mrjana ok now builds are passing :) Let me know if you want me to squash commits before merge.

Now we're now building `dnet` for many archiectures, hopefully we might catch any changes that break compilation

```
linux/amd64...
/go/bin/godep go build -o "bin/dnet-$GOOS-$GOARCH" -tags libnetwork_discovery ./cmd/dnet
linux/386...
/go/bin/godep go build -o "bin/dnet-$GOOS-$GOARCH" -tags libnetwork_discovery ./cmd/dnet
linux/arm...
/go/bin/godep go build -o "bin/dnet-$GOOS-$GOARCH" -tags libnetwork_discovery ./cmd/dnet
windows/amd64...
/go/bin/godep go build -o "bin/dnet-$GOOS-$GOARCH" -tags libnetwork_discovery ./cmd/dnet
windows/386...
/go/bin/godep go build -o "bin/dnet-$GOOS-$GOARCH" -tags libnetwork_discovery ./cmd/dnet
```

**Comment by mrjana** [comment]
LGTM

**Comment by dave-tucker** [comment]
@mavenugo can I get another LGTM?

**Comment by aboch** [comment]
@dave-tucker Can you please rebase

**Comment by mavenugo** [comment]
ping @dave-tucker

**Comment by mrjana** [comment]
ping @dave-tucker Can you please rebase? This time I promise you we will merge it :-)

**Comment by dave-tucker** [comment]
@mrjana @mavenugo 
Now rebased (finally)
Scope has been reduced to just adding the `make cross` target for now. 
Once #770 has merged, we'll be one step closer to adding this to CI

File: test/integration/dnet/helpers.bash [hunk 3]
```diff
@@ -168,7 +168,7 @@ EOF
 	   -v $(pwd)/${TMPC_ROOT}:/scratch \
 	   -v /usr/local/bin/runc:/usr/local/bin/runc \
 	   -w /go/src/github.com/docker/libnetwork \
-	   mrjana/golang ./cmd/dnet/dnet -d -D ${hopt} -c ${tomlfile}
+	   mrjana/golang ./bin/dnet -d -D ${hopt} -c ${tomlfile}
```
Comment (dave-tucker) [inline test/integration/dnet/helpers.bash]: I meant to ask, what's in `mrjana/golang`?
Can we replace with the `libnetwork-build` image?

File: Makefile [hunk 4]
```diff
@@ -1,42 +1,46 @@
-.PHONY: all all-local build build-local check check-code check-format run-tests check-local integration-tests install-deps coveralls circle-ci start-services clean
+.PHONY: all all-local build build-local clean cross cross-local check check-code check-format run-tests integration-tests check-local coveralls circle-ci-cross circle-ci-build circle-ci-check circle-ci
 SHELL=/bin/bash
-build_image=libnetworkbuild
+build_image=libnetwork-build
```
Comment (aboch) [inline Makefile]: I am not sure why #667 removed the dash from the name.
But you may want to check with @LK4D4 before re-adding the `-`.

File: Makefile [hunk 5]
```diff
@@ -1,42 +1,46 @@
...
 build: ${build_image}.created
 	@echo "Building code... "
 	@${docker} ./wrapmake.sh build-local
 	@echo "Done building code"
 
 build-local:
-	@$(shell which godep) go build  ./...
-	@$(shell which godep) go build -o ./cmd/dnet/dnet ./cmd/dnet
+	@mkdir -p "bin"
+	$(shell which godep) go build -o "bin/dnet" ./cmd/dnet
```
Comment (aboch) [inline Makefile]: +1 on moving dnet to `bin/`

**Comment by aboch** [comment]
LGTM

**Pull request merged**